toxcore/crypto_core.c

```diff
@@ -123,53 +123,70 @@ int decrypt_data_symmetric(const uint8_t *secret_key, const uint8_t *nonce, cons
+    int ret = encrypt_data_symmetric(k, nonce, plain, length, encrypted);
+    sodium_memzero(k, sizeof k);
+    return ret;
 }
 
 int decrypt_data(const uint8_t *public_key, const uint8_t *secret_key, const uint8_t *nonce,
                  const uint8_t *encrypted, uint32_t length, uint8_t *plain)
 {
+    if (!public_key || !secret_key)
+        return -1;
+
     uint8_t k[crypto_box_BEFORENMBYTES];
     encrypt_precompute(public_key, secret_key, k);
-    return decrypt_data_symmetric(k, nonce, encrypted, length, plain);
+    int ret = decrypt_data_symmetric(k, nonce, encrypted, length, plain);
+    sodium_memzero(k, sizeof k);
+    return ret;
 }
 
 
 /* Increment the given nonce by 1. */
 void increment_nonce(uint8_t *nonce)
 {
-    uint32_t i;
-
-    for (i = crypto_box_NONCEBYTES; i != 0; --i) {
-        ++nonce[i - 1];
-
-        if (nonce[i - 1] != 0)
-            break;
+    /* FIXME use increment_nonce_number(nonce, 1) or sodium_increment (change to little endian)
+     * NOTE don't use breaks inside this loop
+     * In particular, make sure, as far as possible,
+     * that loop bounds and their potential underflow or overflow
+     * are independent of user-controlled input (you may have heard of the Heartbleed bug).
+     */
+    uint32_t i = crypto_box_NONCEBYTES;
+    uint_fast16_t carry = 1U;
+    for (; i != 0; --i) {
+        carry += (uint_fast16_t) nonce[i - 1];
+        nonce[i - 1] = (uint8_t) carry;
+        carry >>= 8;
     }
 }
 /* increment the given nonce by num */
-void increment_nonce_number(uint8_t *nonce, uint32_t num)
+void increment_nonce_number(uint8_t *nonce, uint32_t host_order_num)
 {
-    uint32_t num1, num2;
-    memcpy(&num1, nonce + (crypto_box_NONCEBYTES - sizeof(num1)), sizeof(num1));
-    num1 = ntohl(num1);
-    num2 = num + num1;
-
-    if (num2 < num1) {
-        uint32_t i;
-
-        for (i = crypto_box_NONCEBYTES - sizeof(num1); i != 0; --i) {
-            ++nonce[i - 1];
-
-            if (nonce[i - 1] != 0)
-                break;
-        }
+    /* NOTE don't use breaks inside this loop
+     * In particular, make sure, as far as possible,
+     * that loop bounds and their potential underflow or overflow
+     * are independent of user-controlled input (you may have heard of the Heartbleed bug).
+     */
+    const uint32_t big_endian_num = htonl(host_order_num);
+    const uint8_t* const num_vec = (const uint8_t*) &big_endian_num;
+    uint8_t num_as_nonce[crypto_box_NONCEBYTES] = {0};
+    num_as_nonce[crypto_box_NONCEBYTES - 4] = num_vec[0];
+    num_as_nonce[crypto_box_NONCEBYTES - 3] = num_vec[1];
+    num_as_nonce[crypto_box_NONCEBYTES - 2] = num_vec[2];
+    num_as_nonce[crypto_box_NONCEBYTES - 1] = num_vec[3];
+
+    uint32_t i = crypto_box_NONCEBYTES;
+    uint_fast16_t carry = 0U;
+    for (; i != 0; --i) {
+        carry += (uint_fast16_t) nonce[i] + (uint_fast16_t) num_as_nonce[i];
+        nonce[i] = (unsigned char) carry;
```

Comment:
@irungentoo thank you. That was a bad bug. I wrote unit tests for that case. See [PR](/kpp/2tox/pull/5).